src/main/java/com/mongodb/spark/sql/connector/connection/MongoConnectionProvider.java

```diff
@@ -41,12 +40,9 @@
  * MongoClientCache}.
  */
 @ApiStatus.Internal
-public final class MongoConnectionProvider implements Serializable {
+@ThreadSafe
+public final class MongoConnectionProvider {
```

Comment:
`MongoConnectionProvider` must be thread-safe, but `MongoConfig` is not thread-safe, and `MongoConnectionProvider.withClient` calls `MongoConfig.getMongoClientFactory`, which is a contradiction.

P.S.
You may say that I proposed to mark `MongoConfig` as not thread-safe in https://github.com/mongodb/mongo-spark/pull/55#discussion_r720911762, but I did that based on its code (as I mentioned) and not its usage. At that time it was unspecified whether `MongoConnectionProvider` must be thread-safe or not, so I could not judge based on the `MongoConfig` usage.

Reply:
Sorry MongoConfig is ThreadSafe.